In [1]:
import pandas as pd
import urllib
import os
import string
import re
import numpy as np

In [2]:
df=pd.read_csv("Data/noms_voies_actuelles_paris.csv", sep=';')
df.head()

DATE ARRET  CDGI  \
0  1864-02-03T00:53:28+00:53  4166   
1                        NaN  4994   
2                        NaN  5441   
3                        NaN  8208   
4  1885-10-11T00:53:28+00:53  9057   

                                     QUARTIER TYPE DE VOIE PREFIXE de VOIE  \
0                                   Roquette.          rue             NaN   
1                             Saint-Ambroise.         cité             NaN   
2                             Saint-Ambroise.          rue             NaN   
3  Folie-Méricourt. Saint-Ambroise. Roquette.    boulevard             NaN   
4                             Saint-Ambroise.          rue             NaN   

      NOM de VOIE              Libellé VOIE CVOIE  \
0         gerbier               rue Gerbier  4112   
1            joly                 cité Joly  4886   
2        léchevin              rue Léchevin  5376   
3  richard lenoir  boulevard Richard Lenoir  8197   
4         spinoza               rue Spinoza  8671   

                                          HISTORIQUE  LARGEUR    LONG  \
0  Orig. du nom. ~ Pierre Jean Baptiste Gerbier (...     13.0   174.0   
1             Orig. du nom. ~ M. Joly, propriétaire.     12.0   135.0   
2  Orig. du nom. ~ M. Léchevin, propriétaire.\r\n...     15.0    73.0   
3  Orig. du nom. ~ François Richard (1765-1839), ...     60.0  1500.0   
4  Orig. du nom. ~ Baruch Spinoza (1632-1677), ph...     12.0    77.0   

                                              DETAIL ARRONDISSEMENT  \
0  Ouverture et alignements (non retenus au POS e...            XIe   
1  Classement, alignements (non retenus au POS et...            XIe   
2  Classement, alignements (non retenus au POS et...            XIe   
3  Alignements (non retenus au POS et non repris ...            XIe   
4  Ouverture, alignements (non retenus au POS et ...            XIe   

  STATUT PRESUME                                              DEBUT  \
0  voie publique                      rue de la Folie-Regnault, 15.   
1  voie publique                           rue du Chemin Vert, 121.   
2  voie publique                             avenue Parmentier, 64.   
3  voie publique  boulevard Beaumarchais, 2 et place de la Basti...   
4  voie publique                      avenue de la République, 103.   

                                                 FIN  
0                       rue de la Roquette, 168 bis.  
1                                         en impasse  
2                     passage Saint-Ambroise, 9 bis.  
3  avenue de la République, 22, rue Rampon et pla...  
4                     boulevard de Ménilmontant, 81.

In [3]:
len(df)

6475

# 1_Cleaning streets names

On va utiliser la variable "HISTORIQUE", pour extraire des informations relatives à la personne qui a donné son nom à la voie.

In [4]:
# Cleaner à la main : ceux erreurs
#df['HISTORIQUE'][25] = "Grace Murray Hopper (New-York, 1906- comté d'Arlington, 1992), mathématicienne et informaticienne américaine, ayant mis au point le premier compilateur ainsi que le langage Cobol. \r\nHistorique. ~ voie créée dans le secteur de la Halle Freyssinet, précédemment voie GR/13. Anciennement partie de la voie GM/13."

In [5]:
# Removing expressions "Origin. du nom"
removed_expressions = ["Orig. du nom. ~ ", "Orig.du nom.~ ", "Orig. Du nom. ~"]
for expression in removed_expressions :
    df["HISTORIQUE"] = df['HISTORIQUE'].astype(str).apply(lambda x: x.replace(expression, ''))

In [6]:
new = df['HISTORIQUE'].str.split("\r\nHistorique. ~ |Historique. ~ ", n = 1, expand = True) 
df['Histoire'] = new[1]
df['Rues'] = new[0]

In [7]:
new = df["Rues"].str.split(";", n = 1, expand = True)
df['Quartier'] = new[1]
df['Rues'] = new[0]

In [8]:
df

DATE ARRET  CDGI  \
0     1864-02-03T00:53:28+00:53  4166   
1                           NaN  4994   
2                           NaN  5441   
3                           NaN  8208   
4     1885-10-11T00:53:28+00:53  9057   
5                           NaN  4865   
6                           NaN  2320   
7     1965-05-05T01:00:00+01:00  6844   
8                           NaN  R530   
9     1911-03-21T01:00:00+01:00  2904   
10                          NaN  X302   
11    1889-01-25T00:53:28+00:53  6484   
12                          NaN  8037   
13    1972-07-04T01:00:00+01:00  2216   
14                          NaN  R700   
15    1985-03-09T01:00:00+01:00  8923   
16    1885-09-12T00:53:28+00:53  3472   
17                          NaN  X158   
18                          NaN  7271   
19    1873-10-11T00:53:28+00:53  9171   
20                          NaN  8910   
21    1867-02-27T00:53:28+00:53  9127   
22    1956-03-13T01:00:00+01:00  6148   
23    1899-07-31T01:00:00+01:00  7139   
24    2017-09-25T02:00:00+02:00   NaN   
25    2017-09-05T02:00:00+02:00  4258   
26                          NaN  7344   
27    1985-02-04T01:00:00+01:00  0373   
28    1999-12-30T01:00:00+01:00  4385   
29                          NaN  9492   
...                         ...   ...   
6445  1996-12-18T01:00:00+01:00  6931   
6446  1864-08-24T00:53:28+00:53  4229   
6447                        NaN  5033   
6448  1905-07-18T01:00:00+01:00  9491   
6449                        NaN  3747   
6450  1937-10-02T01:00:00+01:00  3315   
6451                        NaN  A719   
6452  1993-09-03T02:00:00+02:00  9555   
6453                        NaN  0535   
6454  1930-12-24T01:00:00+01:00  2698   
6455  1899-02-08T01:00:00+01:00  6057   
6456  1984-11-26T01:00:00+01:00  0113   
6457                        NaN  1535   
6458                        NaN  6815   
6459                        NaN  7521   
6460  1877-01-02T00:53:28+00:53  7534   
6461                        NaN  1141   
6462                        NaN  X173   
6463  1975-10-30T01:00:00+01:00  3857   
6464  1934-01-13T01:00:00+01:00  7430   
6465  1972-03-15T01:00:00+01:00  0334   
6466  1907-06-24T01:00:00+01:00  1992   
6467                        NaN  A805   
6468  1894-12-28T01:00:00+01:00  3419   
6469                        NaN  X406   
6470  1951-02-05T01:00:00+01:00  4405   
6471  1885-10-11T00:53:28+00:53  9992   
6472  1997-05-14T02:00:00+02:00  9332   
6473  1864-08-24T00:53:28+00:53  1315   
6474                        NaN  1848   

                                        QUARTIER TYPE DE VOIE PREFIXE de VOIE  \
0                                      Roquette.          rue             NaN   
1                                Saint-Ambroise.         cité             NaN   
2                                Saint-Ambroise.          rue             NaN   
3     Folie-Méricourt. Saint-Ambroise. Roquette.    boulevard             NaN   
4                                Saint-Ambroise.          rue             NaN   
5                                 Quinze-Vingts.          rue             NaN   
6                                 Quinze-Vingts.       avenue              de   
7                                       Bel Air.          rue           de la   
8                                        Picpus.       avenue             des   
9                                        Picpus.       avenue             NaN   
10                                       Picpus.         voie             NaN   
11                                      Bel Air.          rue             NaN   
12                         Bercy. Quinze-Vingts.         quai           de la   
13                                       Picpus.          rue              du   
14                                      Bel Air.        route             des   
15                                    Plaisance.        place              de   
16                               Porte Dauphine.       avenue              d'   
17                  

## 1_1 Distinction noms de personnes

On distingue les rues qui portent des noms d'individus et les autres.

Plusieurs façons de reconnaître les rues qui portent des noms d'individus :
- On a souvent les dates de naissance et mort entre parenthèses pour les rues qui portent des noms de personnes
- S'il s'agit d'un propriétaire (qui possédait donc des immeubles dans cette rue), on peut filter avec "propriétaire"
- Pour certaines rues (plutôt des impasses), on n'a pas la description de la personne, car il s'agit d'une rue au voisinage d'un boulevard par exemple (et l'information sur la personne est disponible à la ligne correspondant au boulevard)
- Ou sinon, si on n'a ni l'un ni l'autre, on peut utiliser la liste des prénoms (puisque pour certains individus, on n'a ni les autres informations ci-dessus)

### 1. 1. 1. Voisinages

In [9]:
# dataframe voisinages: on remplace la "Rues" par la vraie référence à la rue
voisinages = df.loc[df['Rues'].str.startswith("Voisinage")].reset_index()
voisinages["Voisinage rue"] = voisinages["Rues"]
voisinages = pd.merge(voisinages.loc[:, voisinages.columns != 'Rues'], 
                      df.loc[~df['Rues'].str.contains("Voisinage"), ['Rues', 'NOM de VOIE']], 
                      how='left', 
                      on='NOM de VOIE')

In [10]:
# Issue: pour certaines rues, on n'a pas exactement le même Libellé voie entre la rue principale et celle au voisinage. 
# Ou sinon, il est juste inscrit "voisinage", même pour la rue principale (si voisinage d'une église par ex)
# On va corriger ça.
list_voisinages_pb = voisinages.loc[voisinages['Rues'].isna(), "NOM de VOIE"].tolist()

for voisinage in list_voisinages_pb :
    if len(df.loc[df['NOM de VOIE'].str.contains(voisinage)]) == len(df.loc[df['NOM de VOIE'].str.startswith(voisinage)]) :
        voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Rues"] = voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Voisinage rue"]
    
    elif len(df.loc[df['NOM de VOIE'].str.contains(voisinage)])==1:
        voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Rues"] = voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Voisinage rue"]

    elif len(df.loc[df['NOM de VOIE'].str.contains(voisinage)])==2:
        new_historique_rue = df.loc[(~df['Rues'].str.contains("Voisinage")) & (df['NOM de VOIE'].str.contains(voisinage)), 'Rues']
        if new_historique_rue.empty :
            voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Rues"] = voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Voisinage rue"]
        else :
            voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Rues"] = new_historique_rue.values[0]
    
    else :
        voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Rues"] = voisinages.loc[voisinages['NOM de VOIE']==voisinage, "Voisinage rue"]

In [11]:
voisinages = voisinages.drop_duplicates(subset='index', keep="first")
voisinages = voisinages.set_index("index")
voisinages.index.name=None

In [12]:
df = df.merge(voisinages.loc[:, ['Rues']], how='left', left_index=True, right_index=True)

In [13]:
df.loc[~df['Rues_y'].isna(), 'Rues_x'] = df['Rues_y']
df = df.drop(['Rues_y'], axis=1).rename(columns={'Rues_x' : 'Rues'})

### 1. 1. 2. Labeliser: si c'est une personne

In [14]:
# Dates de naissance et de mort + propriétaire
df['Personne'] = 0
df.loc[df['HISTORIQUE'].str.contains("\(|, propriétaire|Nom de propriétaire"), ['Personne']] = 1

In [15]:
# Ajouter les saints (pas toujours ajouté)
df.loc[df['NOM de VOIE'].str.contains("saint"), ['Personne']] = 1

In [16]:
df['Personne'].value_counts()

1    3961
0    2514
Name: Personne, dtype: int64

In [17]:
# On ne les a pas tous. Exemple:
df[df.index==6446]

DATE ARRET  CDGI        QUARTIER TYPE DE VOIE  \
6446  1864-08-24T00:53:28+00:53  4229  Place Vendôme.          rue   

     PREFIXE de VOIE NOM de VOIE  Libellé VOIE CVOIE  \
6446             NaN    gomboust  rue Gomboust  4183   

                                             HISTORIQUE  LARGEUR  LONG  \
6446  Jacques Gomboust, ingénieur du Roi, auteur d'u...     10.0  60.0   

                                                 DETAIL ARRONDISSEMENT  \
6446  Alignements (non retenus au POS et non repris ...            Ier   

     STATUT PRESUME                DEBUT                                FIN  \
6446  voie publique  rue Saint-Roch, 57.  place du Marché Saint-Honoré, 38.   

                                               Histoire  \
6446  Précédemment, rue de la Corderie Saint-Honoré ...   

                                                   Rues  \
6446  Jacques Gomboust, ingénieur du Roi, auteur d'u...   

                                               Quartier  Personne  
6446   voisinage de l'Hôtel du Saint-Esprit où ce pl...         0

# 2. Ajouter genre

In [18]:
for col in ['NOM de VOIE', 'HISTORIQUE', 'Rues']:
    df[col] = df[col].str.lower()

In [19]:
names = pd.read_csv("Data/Dico_gender_proportion.csv").drop('index', axis=1)
prenoms_list = names['prenoms'].tolist()

In [20]:
df['Sexe'] = float('nan')
for prenom in prenoms_list :
    sexe = names.loc[names['prenoms']==prenom, 'sexe'].values[0]
    df.loc[(df['Personne']==1) & (df['Rues'].str.contains(prenom)), 'Sexe'] = sexe
    df.loc[(df['Personne']==1) & (df['HISTORIQUE'].str.contains(prenom) & df['Sexe'].isna()), 'Sexe'] = sexe

# Pour les propriétaires :
df.loc[(df['Personne']==1) & (df['HISTORIQUE'].str.startswith('m.') & df['HISTORIQUE'].str.contains('propriétaire')), 'Sexe'] = 'M'
df.loc[(df['Personne']==1) & (df['HISTORIQUE'].str.contains('nom de propriétaire')), 'Sexe'] = 'M'

# Les saints :
df.loc[(df['Personne']==1) & (df['HISTORIQUE'].str.contains('sainte')), 'Sexe'] = 'F'
df.loc[(df['Personne']==1) & (df['HISTORIQUE'].str.contains('saint') & df['Sexe'].isna()), 'Sexe'] = 'M'

# Autres
df.loc[df['Personne']==0 , 'Sexe'] = 'Autre'

In [21]:
df['Sexe'].value_counts()

M        2890
Autre    2514
F        1001
Name: Sexe, dtype: int64

In [22]:
df

DATE ARRET  CDGI  \
0     1864-02-03T00:53:28+00:53  4166   
1                           NaN  4994   
2                           NaN  5441   
3                           NaN  8208   
4     1885-10-11T00:53:28+00:53  9057   
5                           NaN  4865   
6                           NaN  2320   
7     1965-05-05T01:00:00+01:00  6844   
8                           NaN  R530   
9     1911-03-21T01:00:00+01:00  2904   
10                          NaN  X302   
11    1889-01-25T00:53:28+00:53  6484   
12                          NaN  8037   
13    1972-07-04T01:00:00+01:00  2216   
14                          NaN  R700   
15    1985-03-09T01:00:00+01:00  8923   
16    1885-09-12T00:53:28+00:53  3472   
17                          NaN  X158   
18                          NaN  7271   
19    1873-10-11T00:53:28+00:53  9171   
20                          NaN  8910   
21    1867-02-27T00:53:28+00:53  9127   
22    1956-03-13T01:00:00+01:00  6148   
23    1899-07-31T01:00:00+01:00  7139   
24    2017-09-25T02:00:00+02:00   NaN   
25    2017-09-05T02:00:00+02:00  4258   
26                          NaN  7344   
27    1985-02-04T01:00:00+01:00  0373   
28    1999-12-30T01:00:00+01:00  4385   
29                          NaN  9492   
...                         ...   ...   
6445  1996-12-18T01:00:00+01:00  6931   
6446  1864-08-24T00:53:28+00:53  4229   
6447                        NaN  5033   
6448  1905-07-18T01:00:00+01:00  9491   
6449                        NaN  3747   
6450  1937-10-02T01:00:00+01:00  3315   
6451                        NaN  A719   
6452  1993-09-03T02:00:00+02:00  9555   
6453                        NaN  0535   
6454  1930-12-24T01:00:00+01:00  2698   
6455  1899-02-08T01:00:00+01:00  6057   
6456  1984-11-26T01:00:00+01:00  0113   
6457                        NaN  1535   
6458                        NaN  6815   
6459                        NaN  7521   
6460  1877-01-02T00:53:28+00:53  7534   
6461                        NaN  1141   
6462                        NaN  X173   
6463  1975-10-30T01:00:00+01:00  3857   
6464  1934-01-13T01:00:00+01:00  7430   
6465  1972-03-15T01:00:00+01:00  0334   
6466  1907-06-24T01:00:00+01:00  1992   
6467                        NaN  A805   
6468  1894-12-28T01:00:00+01:00  3419   
6469                        NaN  X406   
6470  1951-02-05T01:00:00+01:00  4405   
6471  1885-10-11T00:53:28+00:53  9992   
6472  1997-05-14T02:00:00+02:00  9332   
6473  1864-08-24T00:53:28+00:53  1315   
6474                        NaN  1848   

                                        QUARTIER TYPE DE VOIE PREFIXE de VOIE  \
0                                      Roquette.          rue             NaN   
1                                Saint-Ambroise.         cité             NaN   
2                                Saint-Ambroise.          rue             NaN   
3     Folie-Méricourt. Saint-Ambroise. Roquette.    boulevard             NaN   
4                                Saint-Ambroise.          rue             NaN   
5                                 Quinze-Vingts.          rue             NaN   
6                                 Quinze-Vingts.       avenue              de   
7                                       Bel Air.          rue           de la   
8                                        Picpus.       avenue             des   
9                                        Picpus.       avenue             NaN   
10                                       Picpus.         voie             NaN   
11                                      Bel Air.          rue             NaN   
12                         Bercy. Quinze-Vingts.         quai           de la   
13                                       Picpus.          rue              du   
14                                      Bel Air.        route             des   
15                                    Plaisance.        place              de   
16                               Porte Dauphine.       avenue              d'   
17                  

In [23]:
pd.crosstab(df['TYPE DE VOIE'], df['Sexe'])

Sexe                Autre    F     M
TYPE DE VOIE                        
allée                  44   26    52
arcades                 1    0     0
autoroute               2    0     0
avenue                 80   84   184
belvédère               0    0     1
boulevard              23   26    66
carrefour              14    1     0
chaussée                2    0     0
chemin                 14    0     0
cité                   64    3    37
cour                   28    8    15
cours                   3    0     2
esplanade               3    5    14
galerie                19    2     1
grande avenue           1    0     0
hameau                  5    0     2
impasse               105   16    67
jardin                  0    1     0
parvis                  3    1     3
parvis, place           0    0     1
passage               171   28    93
passage souterrain      0    0     1
passerelle              1    0     4
patio-place             0    0     1
place                 156  112   292
placette                0    1     3
pont                   20    4     6
port                   22    1     3
porte                  15    3     0
promenade               2    4    13
péristyle               2    0     0
quai                   23   10    21
rond-point              1    1     2
route                  81    5     0
rue                   884  632  1899
ruelle                  4    1     2
sente                   2    0     0
sentier                 6    1     0
square                115    6    44
terrasse                3    1     2
villa                 139   16    57
voie                  447    1     2

# Suite

A corriger :
- Certaines rues ne sont pas labelisées comme personnes (ex : 6446)
- Certaines rues le sont alors qu'elles ne sont pas des personnes (ex : 6458, il faut bien corriger pour parenthèses + chiffres) ; aussi bd des italiens.
- Dans le genre, pb si prénom étrange (comme Spinoza) : utiliser liste des célébrités
- Certaines erreurs, quand des femmes portent un nom de famille d'homme (ex : 24, car place Marcelle Henry). IL faut modifier l'attribution des noms: looper pour distinguer nom et prénom (avec la position dans le texte peut-être). L'inverse aussi (4986, Garibaldi labelisé comme une femme).

Pour les voisinages : comme on cherche maintenant dans "HISTORIQUE". Il faut remodifier pour chercher dans "Rues" quand c'est un voisinage (créer une autre variable).
Autre idée: récupérer via Wikipédia (liste des rues portant des noms de femmes.

In [24]:
# Toujours pas parfait du tout :
df.loc[(df['TYPE DE VOIE']=="boulevard") & (df['Sexe']=='F')] 

DATE ARRET  CDGI  \
230   1864-08-19T00:53:28+00:53  7006   
471   1873-10-11T00:53:28+00:53  9623   
684   1931-04-21T01:00:00+01:00  9982   
695   1864-08-19T00:53:28+00:53  8690   
770   1855-09-25T00:53:28+00:53  8894   
1028  1931-04-21T01:00:00+01:00  2809   
1185  1870-10-25T00:53:28+00:53  9907   
1221                        NaN  4758   
1531                        NaN  8619   
1801                        NaN  1557   
1889  1911-02-20T01:00:00+01:00  5081   
1982  1931-04-21T01:00:00+01:00  3741   
2259  1885-10-11T00:53:28+00:53  3957   
2353  1931-04-21T01:00:00+01:00  8108   
2423                        NaN  1097   
2786  1987-09-01T02:00:00+02:00  4070   
2860  1946-08-06T02:00:00+02:00  0069   
2971  1864-12-30T00:53:28+00:53  4310   
3285  1931-04-21T01:00:00+01:00  9019   
3324  1931-04-21T01:00:00+01:00  2917   
3983  1932-01-22T01:00:00+01:00  1070   
4096  1931-08-20T01:00:00+01:00  7298   
4642                        NaN  5907   
4962                        NaN  5917   
4986  1864-02-03T00:53:28+00:53  1333   
5621  1930-11-08T01:00:00+01:00  0580   

                                               QUARTIER TYPE DE VOIE  \
230              Saint-Germain l'Auxerrois. Notre-Dame.    boulevard   
471                                             Necker.    boulevard   
684                                             Ternes.    boulevard   
695                    Jardin des Plantes. Salpêtrière.    boulevard   
770   Saint-Germain l'Auxerrois. Halles. Bonne Nouve...    boulevard   
1028                                            Ternes.    boulevard   
1185  Folie-Méricourt. Saint-Ambroise. Roquette. Sai...    boulevard   
1221  Gaillon. Vivienne. Chaussée d'Antin. Faubourg ...    boulevard   
1531  Saint-Victor. Sorbonne. Monnaie. Odéon. Saint-...    boulevard   
1801                                           Bel Air.    boulevard   
1889                                   Folie-Méricourt.    boulevard   
1982                                       Batignolles.    boulevard   
2259                                            Necker.    boulevard   
2353                                Plaine de Monceaux.    boulevard   
2423                 Bonne Nouvelle. Porte Saint-Denis.    boulevard   
2786                                             Javel.    boulevard   
2860                                         Plaisance.    boulevard   
2971                                          Grenelle.    boulevard   
3285                                            Ternes.    boulevard   
3324                                       Batignolles.    boulevard   
3983                                         Epinettes.    boulevard   
4096                                            Ternes.    boulevard   
4642  Rochechouart. Saint-Vincent de Paul. Porte Sai...    boulevard   
4962            Sur le territoire de Neuilly-sur-Seine.    boulevard   
4986                        Plaisance. Petit Montrouge.    boulevard   
5621                                            Ternes.    boulevard   

     PREFIXE de VOIE            NOM de VOIE  \
230               du                 palais   
471               de              vaugirard   
684            de l'                   yser   
695              NaN           saint-marcel   
770               de             sébastopol   
1028              de                dixmude   
1185             NaN               voltaire   
1221             des               italiens   
1531             NaN          saint-germain   
1801             NaN                 carnot   
1889             NaN            jules ferry   
1982              du           fort de vaux   
2259             NaN              garibaldi   
2353              de                  reims   
2423              de         bonne nouvelle   
2786              du  général martial valin   
2860             NaN         adolphe pinard   
2971              de               grenelle   
3285           de la                  somme   
3324              de        